File: C-code/FactTable.c
#include<stdio.h>
#include <string.h>
#include "FactTable.h"
#include <stdlib.h>

struct _FactNode {
	char* fact;
	char* category;
	struct FactNode* next;
};

struct _CategoryNode {
	char* category;
	int count;
	struct CategoryNode* next;
};

FactNode* hashtable[100];
CategoryNode * counts;

int _hash(char* category) {
	int hasher = 0;
	char* c = category;
	while (*c != '\0') {
		hasher += *c;
		c++;
	}
	hasher = hasher%100;
	return hasher;
}
void hashPut(FactNode * fact)
{
		int hasher = _hash(fact->category);
		fact->next = hashtable[hasher];
		hashtable[hasher]=fact;
		char * cat = fact->category;
		CategoryNode * curr = counts;
		while(1)
		{
			if(strcmp(curr->category,fact->category)==0)
			{
				curr->count=curr->count+1;
				return;
			}
			if(curr->next!=NULL)
				curr = curr->next;
			else
				break;
		}
		CategoryNode * node= malloc(sizeof(CategoryNode *));
		node->category=fact->category;
		node->count = 1;
		node->next = NULL;
		curr->next= node;
}
FactNode * hashGet(char * category)
{
	int numFacts = 0;
	CategoryNode * curr = counts;
	while(curr!=NULL)
	{
		if(strcmp(curr->category,category)==0)
			{
				numFacts = curr->count;
				break;
			}
			curr = curr->next;
	}
	if(numFacts==0)
	{
		printf("Error, category not found in arraylist CategoryCount\n");
		return NULL;
	}
	int rando = rand() % numFacts+1;
	int hasher = hash(category);
	FactNode * curren = hashtable[hasher];
	while(1)
	{
		if(strcmp(curren->category,category)==0)
			rando=rando-1;
		if(rando ==0)
			return curren;
		curren = curren->next;
	}
	return NULL;
}

main() {

	FactNode** hashtable = malloc(sizeof(FactNode **));
	CategoryNode * counts = malloc(sizeof(CategoryNode *));
	
}

File: FactTable.java
import java.util.ArrayList;
public class FactTable {
	private FactNode[] Hashtable;
	private ArrayList<CategoryCount> counts; // The number of facts in each category will help randomize the output
	public FactTable()
	{
		Hashtable= new FactNode[100];
		counts = new ArrayList<CategoryCount>();
	}
	private static int hash(String category)//very predictable, we don't need sporadic behavior. Sparse enough.
	{
		int hasher=0;
		for(int i = 0; i<category.length();i++)
		{
			hasher+=(int) category.charAt(i);
		}
		hasher = hasher %100;
		return hasher;
	}
	public void put(FactNode fact)
	{
		int hasher = hash(fact.getCategory());
		fact.next = Hashtable[hasher];
		String cat = fact.getCategory();
		for(int i = 0; i< counts.size(); i++)//Increment the count of facts in the category.
		{
			if(counts.get(i).getCategory().equals(fact.getCategory()))
			{
				counts.get(i).increment();
				return;
			}
		}
		counts.add(new CategoryCount(cat,1));//In case not there, make new category
	}
	public FactNode get(String category)
	{
		int numFacts = 0;
		for(int i = 0; i<counts.size();i++)//retrieve the number of facts in the category
		{
			if(counts.get(i).getCategory().equals(category));
			{
				numFacts = counts.get(i).getCount();
				break;
			}
		}
		if(numFacts==0)
		{
			System.out.println("Error, category not found in arraylist CategoryCount");
			return null;
		}
		int rand = (int)(Math.random()*numFacts)+1;
		int hasher = hash(category);
		FactNode curr = Hashtable[hasher];
		while(true)//get the rand-th fact in that category
		{
			if(curr.getCategory()==category)
				rand--;
			if(rand ==0)
				return curr;
			curr = curr.next;
		}
	}
	
	

}
